### PR TITLE
Add support for application hook.

### LINKED ISSUE #141: [RFC] Add authentication hook for application
I was thinking how the authentication could be configured in a more clean way. Currently we suggest to configure authentication in middleware hook, but the method could become too large and harder to maintain as the application grows.

One of the options is to have a dedicated application hook similar to the `console`, `middleware` or `routes`.

This is how it could look like:

```php
    public function authentication($authentication)
    {
        $authentication->loadIdentifier('Authentication.Password');
        $authentication->loadAuthenticator('Authentication.Session');
        $authentication->loadAuthenticator('Authentication.Form');

        return $authentication;
    }

    public function middleware($middleware)
    {
        $middleware
            ->add(ErrorHandlerMiddleware::class)
            ->add(AssetMiddleware::class)
            ->add(new RoutingMiddleware($this))
            ->add(new AuthenticationMiddleware($this));

        return $middleware;
    }
```

The `AuthenticationService` could be configured in a hook and the application instance could be passed to the `AuthenticationMiddleware`.

Sometimes different authentication is needed for different scopes. This could be achieved with named authentication services.

```php
    public function authentication($authentication)
    {
        $authentication->loadIdentifier('Authentication.Password');
        $authentication->loadAuthenticator('Authentication.Session');
        $authentication->loadAuthenticator('Authentication.Form');

        return $authentication;
    }

    public function authenticationApi($authentication)
    {
        $authentication->loadIdentifier('Authentication.Token');
        $authentication->loadAuthenticator('Authentication.Token');

        return $authentication;
    }

    public function routes($routes)
    {
        $routes->registerMiddleware('auth', new AuthenticationMiddleware($this));
        $routes->registerMiddleware('apiAuth', new AuthenticationMiddleware($this, 'api'));

        $routes->scope('/', function (RouteBuilder $routes) {
            $routes->applyMiddleware('auth');
        });

        $routes->scope('/api', function (RouteBuilder $routes) {
            $routes->applyMiddleware('apiAuth');
        });
    }

    public function middleware($middleware)
    {
        $middleware
            ->add(ErrorHandlerMiddleware::class)
            ->add(AssetMiddleware::class)
            ->add(new RoutingMiddleware($this));

        return $middleware;
    }
```

I think this looks cleaner and better structured than having all configuration in one method.

### PR DESCRIPTION
**Pull request by robertpustulka**
Fixes #141

### CODE REVIEW

**Comment by robertpustulka** [comment]
I'm wondering if request (and response?) objects should be passed to the authentication hook?

I'm developing an app where authentication depends on a request. I have achieved this using the intermediate middleware, but maybe there are more use cases where request access in the authentication hook would be helpful?

Ie
```php
    public function authentication($authentication, $request)
    {
        $authentication->loadIdentifier('CustomIdentifier', [
            'scope' => $request->getAttribute('scope')
        ]);
        $authentication->loadAuthenticator('Authentication.Session');
        $authentication->loadAuthenticator('Authentication.Form');

        return $authentication;
    }
```

File: src/Middleware/AuthenticationMiddleware.php [hunk 1]
```diff
@@ -65,10 +90,37 @@ public function __invoke(ServerRequestInterface $request, ResponseInterface $res
...
+        if ($this->subject instanceof AuthenticationServiceInterface) {
+            return $this->subject;
+        }
+
+        $method = 'authentication' . ucfirst($this->name);
+        if (!method_exists($this->subject, $method)) {
+            if (strlen($this->name)) {
+                $message = sprintf('Method `%s` for `%s` authentication service has not been defined in your `Application` class.', $method, $this->name);
+            } else {
+                $message = sprintf('Method `%s` has not been defined in your `Application` class.', $method);
```
Comment (markstory) [inline src/Middleware/AuthenticationMiddleware.php]: Nice error messages :+1:

**Comment by robertpustulka** [comment]
I added the support for request and response in the hook. Those arguments are optional.

**Comment by robertpustulka** [comment]
Is it good to go?

**Pull request merged**